algorithm/evocoin.c

```diff
@@ -0,0 +1,346 @@
+* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+* OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+* SUCH DAMAGE.
+*
+* This file was originally written by Colin Percival as part of the Tarsnap
+* online backup system.
+*/
+
+#include "config.h"
+#include "miner.h"
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+//#include <algorithm.h>
+
+#include "sph/sph_blake.h"
+#include "sph/sph_bmw.h"
+#include "sph/sph_groestl.h"
+#include "sph/sph_jh.h"
+#include "sph/sph_keccak.h"
+#include "sph/sph_skein.h"
+#include "sph/sph_luffa.h"
+#include "sph/sph_cubehash.h"
+#include "sph/sph_shavite.h"
+#include "sph/sph_simd.h"
+#include "sph/sph_echo.h"
+
+#define INITIAL_DATE 1462060800
+#define HASH_FUNC_COUNT 11
+#define DEFAULT_NTIME "00000000"
+
+
+/* Move init out of loop, so init once externally, and then use one single memcpy with that bigger memory block */
+typedef struct {
+	sph_blake512_context    blake1;
+	sph_bmw512_context      bmw1;
+	sph_groestl512_context  groestl1;
+	sph_skein512_context    skein1;
+	sph_jh512_context       jh1;
+	sph_keccak512_context   keccak1;
+	sph_luffa512_context    luffa1;
+	sph_cubehash512_context cubehash1;
+	sph_shavite512_context  shavite1;
+	sph_simd512_context     simd1;
+	sph_echo512_context     echo1;
+} Xhash_context_holder;
+
+static Xhash_context_holder base_contexts;
+
+
+static void init_Xhash_contexts()
+{
+	sph_blake512_init(&base_contexts.blake1);
+	sph_bmw512_init(&base_contexts.bmw1);
+	sph_groestl512_init(&base_contexts.groestl1);
+	sph_skein512_init(&base_contexts.skein1);
+	sph_jh512_init(&base_contexts.jh1);
+	sph_keccak512_init(&base_contexts.keccak1);
+	sph_luffa512_init(&base_contexts.luffa1);
+	sph_cubehash512_init(&base_contexts.cubehash1);
+	sph_shavite512_init(&base_contexts.shavite1);
+	sph_simd512_init(&base_contexts.simd1);
+	sph_echo512_init(&base_contexts.echo1);
+}
+
+
+uint32_t getCurrentAlgoSeq(uint32_t current_time, uint32_t base_time) {
+	return (current_time - base_time) / (60 * 60 * 24);
+}
+
+void swap(uint8_t *a, uint8_t *b) {
+	uint8_t __tmp = *a;
+	*a = *b;
+	*b = __tmp;
+}
+
+void initPerm(uint8_t n[], uint8_t count) {
+	int i;
+	for (i = 0; i<count; i++)
+		n[i] = i;
+}
+
+int nextPerm(uint8_t n[], uint32_t count) {
+	uint32_t tail, i, j;
+
+	if (count <= 1)
+		return 0;
+
+	for (i = count - 1; i>0 && n[i - 1] >= n[i]; i--);
+	tail = i;
+
+	if (tail > 0) {
+		for (j = count - 1; j>tail && n[j] <= n[tail - 1]; j--);
+		swap(&n[tail - 1], &n[j]);
+	}
+
+	for (i = tail, j = count - 1; i<j; i++, j--)
+		swap(&n[i], &n[j]);
+
+	return (tail != 0);
+}
+
+
+void getAlgoString(char *str, uint32_t count)
+{
+	uint8_t algoList[HASH_FUNC_COUNT];
+	char s[100];
+	char *sptr;
+
+	initPerm(algoList, HASH_FUNC_COUNT);
+
+	int j;
+
+	int k;
+	for (k = 0; k < count; k++) {
+		nextPerm(algoList, HASH_FUNC_COUNT);
+	}
+
+	sptr = str;
```

Comment:
it doesnt make sense to convert int to char string, and later "convert back" to int

Reply:
I'm not sure what are you suggesting here, what int conversion are you talking about?

Reply:
nextPerm return the permutation list into algoList as array of uint8_t,
this array is converted to a human readable string, but is is useless. hash function can use (at line ~183) the integers directly